src/bin/casr-afl.rs

```diff
@@ -0,0 +1,274 @@
+    /// Target command line args.
+    pub target_args: Vec<String>,
+    /// Stdin.
+    pub is_stdin: bool,
+    /// ASAN.
+    pub is_asan: bool,
+}
+
+fn main() -> error::Result<()> {
+    let matches = App::new("casr-afl")
+        .version("2.1.1")
+        .author("Andrey Fedotov  <fedotoff@ispras.ru>, Alexey Vishnyakov <vishnya@ispras.ru>, Georgy Savidov <avgor46@ispras.ru>")
+        .about("Triage crashes found by AFL++")
+        .term_width(90)
+        .arg(
+            Arg::new("log-level")
+                .long("log-level")
+                .short('l')
+                .takes_value(true)
+                .default_value("info")
+                .possible_values(["info", "debug"])
+                .help("Logging level")
+        )
+        .arg(
+            Arg::new("input")
+                .short('i')
+                .long("input")
+                .takes_value(true)
+                .value_name("INPUT_DIR")
+                .required(true)
+                .help("AFL++ work directory")
+                .validator(|arg| {
+                    let i_dir = Path::new(arg);
+                    if !i_dir.exists() {
+                        bail!("Input directory doesn't exists.");
+                    }
+                    if !i_dir.is_dir() {
+                        bail!("Input path should be an AFL++ work directory.");
+                    }
+                    Ok(())
+                })
+        )
+        .arg(
+            Arg::new("output")
+                .short('o')
+                .long("output")
+                .takes_value(true)
+                .required(true)
+                .value_name("OUTPUT_DIR")
+                .help("Output directory with triaged reports")
+        )
+        .arg(
+            Arg::new("no-cluster")
+                .long("no-cluster")
+                .help("Do not cluster CASR reports")
+        )
+        .get_matches();
+
+    // Init log.
+    let log_level = if matches.value_of("log-level").unwrap() == "debug" {
+        LevelFilter::Debug
+    } else {
+        LevelFilter::Info
+    };
+    let _ = TermLogger::init(
+        log_level,
+        ConfigBuilder::new().set_time_to_local(true).build(),
+        TerminalMode::Mixed,
+    );
+
+    // Get all crashes.
+    let mut crashes: Vec<AflCrashInfo> = Vec::new();
+    for node_dir in fs::read_dir(matches.value_of("input").unwrap())? {
+        let path = node_dir?.path();
+        if !path.is_dir() {
+            continue;
+        }
+
+        // Get crashes from one node.
+        let mut crash_info = AflCrashInfo::default();
+        let cmdline_path = path.join("cmdline");
+        if let Ok(cmdline) = fs::read_to_string(&cmdline_path) {
+            crash_info.target_args = cmdline.split_whitespace().map(|s| s.to_string()).collect();
+            crash_info.is_stdin = !cmdline.contains("@@");
+
+            // Check for ASAN symbols.
+            if let Some(target) = crash_info.target_args.first() {
```

Comment:
Maybe we should create function and place it to util?

Reply:
Maybe later this is one place only. I think, it's OK for now.